ocaml-lsp-server/src/ocaml_lsp_server.ml

```diff
@@ -125,33 +125,54 @@ let set_diagnostics rpc doc =
-                  in
-                  let relatedInformation =
-                    match error.sub with
-                    | [] -> None
-                    | _ :: _ ->
-                      Some
-                        (List.map error.sub ~f:(fun (sub : Loc.msg) ->
-                             let location =
-                               let range = Range.of_loc sub.loc in
-                               Location.create ~range ~uri
-                             in
-                             let message = message Loc.print_sub_msg sub in
-                             DiagnosticRelatedInformation.create ~location
-                               ~message))
-                  in
-                  let message = message Loc.print_main error in
-                  create_diagnostic ?relatedInformation range message ~severity))
+              let merlin_diagnostics =
+                List.map errors ~f:(fun (error : Loc.error) ->
+                    let loc = Loc.loc_of_report error in
+                    let range = Range.of_loc loc in
+                    let severity =
+                      match error.source with
+                      | Warning -> DiagnosticSeverity.Warning
+                      | _ -> DiagnosticSeverity.Error
+                    in
+                    let message ppf m =
+                      String.trim (Format.asprintf "%a@." ppf m)
+                    in
+                    let relatedInformation =
+                      match error.sub with
+                      | [] -> None
+                      | _ :: _ ->
+                        Some
+                          (List.map error.sub ~f:(fun (sub : Loc.msg) ->
+                               let location =
+                                 let range = Range.of_loc sub.loc in
+                                 Location.create ~range ~uri
+                               in
+                               let message = message Loc.print_sub_msg sub in
+                               DiagnosticRelatedInformation.create ~location
+                                 ~message))
+                    in
+                    let message = message Loc.print_main error in
+                    create_diagnostic ?relatedInformation range message
+                      ~severity)
+              in
+              let holes_as_err_diags =
+                Query_commands.dispatch pipeline Holes
+                |> List.map ~f:(fun (loc, _type) ->
+                       let range = Range.of_loc loc in
+                       let severity = DiagnosticSeverity.Error in
+                       let message =
+                         "This is a typed hole.\n\
+                          If you meant to use a wildcard pattern, here an \
+                          expression is expected not a pattern. Your code will \
+                          not compile if you do not replace this with a valid \
+                          expression."
```

Comment:
Is it possible to include the type of the expression needed by the hole? Also, the error is a bit long in my eyes. I'd prefer something to the tune of "Typed hole awaiting an expression of type ..."`

Reply:
![image](https://user-images.githubusercontent.com/16353531/123310553-a2727c80-d53f-11eb-81c6-57ba01f3f5b2.png)

Does this look better?

Reply:
Looks much better now. Perhaps we should replace `typed hole` with just `hole`? There are no untyped holes, so the qualifier seems redundant.

Reply:
Great. I am just concerned that new (and perhaps even seasoned) ocamlers may not know what a "hole" is, so I made it "typed hole" to be more googlable & slightly more formal (if that makes sense). So if there are no strong objections, I

Reply:
That's fine with me.